src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs

```diff
@@ -15,24 +15,138 @@
+        private const int BUFFERSIZE = 4 * 5120;
         private const string EventSourceNamePrefix = "OpenTelemetry-";
         private readonly EventLevel logLevel;
         private readonly SelfDiagnosticsConfigRefresher configRefresher;
+        private readonly ThreadLocal<byte[]> writeBuffer = new ThreadLocal<byte[]>(() => null);
 
         public SelfDiagnosticsEventListener(EventLevel logLevel, SelfDiagnosticsConfigRefresher configRefresher)
         {
             this.logLevel = logLevel;
-            this.configRefresher = configRefresher;
+            this.configRefresher = configRefresher ?? throw new ArgumentNullException(nameof(configRefresher));
+        }
+
+        /// <summary>
+        /// Encode a string into the designated position in a buffer of bytes, which will be written as log.
+        /// If isParameter is true, wrap "{}" around the string.
+        /// The buffer should not be filled to full, leaving at least one byte empty space to fill a '\n' later.
+        /// If the buffer cannot hold all characters, truncate the string and replace extra content with "...".
+        /// The buffer is not guaranteed to be filled until the last byte due to variable encoding length of UTF-8,
+        /// in order to prioritize speed over space.
+        /// </summary>
+        /// <param name="str">The string to be encoded.</param>
+        /// <param name="isParameter">Whether the string is a parameter. If true, "{}" will be wrapped around the string.</param>
+        /// <param name="buffer">The byte array to contain the resulting sequence of bytes.</param>
+        /// <param name="position">The position at which to start writing the resulting sequence of bytes.</param>
+        /// <returns>The position of the buffer after the last byte of the resulting sequence.</returns>
+        internal static int EncodeInBuffer(string str, bool isParameter, byte[] buffer, int position)
+        {
+            int charCount = str.Length;
+            int ellipses = isParameter ? "{...}\n".Length : "...\n".Length;
+
+            // Ensure there is space for "{...}\n" or "...\n".
+            if (buffer.Length - position - ellipses < 0)
+            {
+                return position;
+            }
+
+            int estimateOfCharacters = (buffer.Length - position - ellipses) / 2;
+
+            // Ensure the UTF-16 encoded string can fit in buffer UTF-8 encoding.
+            // And leave space for "{...}\n" or "...\n".
+            if (charCount > estimateOfCharacters)
+            {
+                charCount = estimateOfCharacters;
+            }
+
+            if (isParameter)
+            {
+                buffer[position++] = (byte)'{';
+            }
+
+            position += Encoding.UTF8.GetBytes(str, 0, charCount, buffer, position);
+            if (charCount != str.Length)
+            {
+                buffer[position++] = (byte)'.';
```

Comment:
Just wondering if maybe a span copy would be faster here? Probably some bounds-checking we could avoid.

Reply:
I wanted to use the `Span<T> struct` in some other code, but found out that it's [not available in .NET Standard 2.0](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-3.1&viewFallbackFrom=netstandard-2.0).

Reply:
What if we add a netstandard2.1 target and then use a preprocessor directive to target the more efficient version for those with a compatible runtime?

Reply:
What we could do as well is install the https://www.nuget.org/packages/System.Memory/ for netstandard2.0 only and use span everywhere. So we wouldn't need to make preprocessor conditions

Reply:
> 
> 
> What we could do as well is install the https://www.nuget.org/packages/System.Memory/ for netstandard2.0 only and use span everywhere. So we wouldn't need to make preprocessor conditions

I found it's used by `System.Diagnostics.DiagnosticSource`, which is already installed.

Reply:
I did a quick test. Manipulating array itself is the fastest. `Span.Fill` is slightly faster. `Span.CopyTo` is relatively slow.

```csharp
using System;
using System.Diagnostics;

namespace SpeedTest
{
    class Program
    {
        const int COUNT = 1000;
        private readonly byte[] buffer1 = new byte[20];
        private readonly byte[] buffer2 = new byte[20];

        void TestArray()
        {
            Stopwatch sw = Stopwatch.StartNew();
            sw.Start();
            for (int i = 0; i < COUNT; i++)
            {
                int start = 10;
                buffer1[start++] = (byte)'.';
                buffer1[start++] = (byte)'.';
                buffer1[start++] = (byte)'.';
            }
            sw.Stop();
            Console.WriteLine("Array assignment: {0:N0} ticks", sw.ElapsedTicks);
        }
        
        void TestSpan()
        {
            Stopwatch sw = Stopwatch.StartNew();
            Span<byte> dots = stackalloc byte[3];
            dots.Fill((byte)'.');
            int start = 10;
            Span<byte> span = new Span<byte>(buffer2, start, 3);
            sw.Start();
            for (int i = 0; i < COUNT; i++)
            {
                dots.CopyTo(span);
                start += 3;
            }
            sw.Stop();
            Console.WriteLine("Span CopyTo: {0:N0} ticks", sw.ElapsedTicks);
        }

        void TestSpanFill()
        {
            Stopwatch sw = Stopwatch.StartNew();
            int start = 10;
            sw.Start();
            for (int i = 0; i < COUNT; i++)
            {
                Span<byte> span = new Span<byte>(buffer2, 10, 3);
                span.Fill((byte)'.');
                start += 3;
            }
            sw.Stop();
            Console.WriteLine("Span Fill: {0:N0} ticks", sw.ElapsedTicks);
        }

        static void Main(string[] args)
        {
            Program p = new Program();
            for (int i = 0; i < 5; i++)
            {
                p.TestSpan();
                p.TestSpanFill();
                p.TestArray();
                Console.WriteLine();
            }
        }
    }
}
```

```
Span CopyTo: 518 ticks
Span Fill: 165 ticks
Array assignment: 131 ticks

Span CopyTo: 254 ticks
Span Fill: 185 ticks
Array assignment: 128 ticks

Span CopyTo: 259 ticks
Span Fill: 193 ticks
Array assignment: 125 ticks

Span CopyTo: 254 ticks
Span Fill: 195 ticks
Array assignment: 125 ticks

Span CopyTo: 257 ticks
Span Fill: 107 ticks
Array assignment: 88 ticks
```

Reply:
So not gonna change the existing approach unless there is a faster one.

Reply:
Makes sense thanks for trying @xiang17!